src/main/java/com/aws/greengrass/util/platforms/unix/linux/LinuxSystemResourceController.java

```diff
@@ -0,0 +1,174 @@
+            }
+            if (resourceLimit.containsKey(MEMORY_KEY)) {
+                long memoryLimitInKB = (long) resourceLimit.get(MEMORY_KEY);
+
+                // TODO: add input validation
+                String memoryLimit = Long.toString(memoryLimitInKB * 1024);
+                Files.write(Cgroup.Memory.getComponentMemoryLimitPath(component.getServiceName()),
+                        memoryLimit.getBytes(StandardCharsets.UTF_8));
+            }
+
+            if (!Files.exists(Cgroup.CPU.getSubsystemComponentPath(component.getServiceName()))) {
+                initializeCgroup(component, Cgroup.CPU);
+            }
+            if (resourceLimit.containsKey(CPU_KEY)) {
+                double cpu = (double) resourceLimit.get(CPU_KEY);
+
+                byte[] content = Files.readAllBytes(
+                        Cgroup.CPU.getComponentCpuPeriodPath(component.getServiceName()));
+                int cpuPeriodUs = Integer.parseInt(new String(content, StandardCharsets.UTF_8).trim());
+
+                int cpuQuotaUs = (int) (cpuPeriodUs * cpu);
+                String cpuQuotaUsStr = Integer.toString(cpuQuotaUs);
+
+                Files.write(Cgroup.CPU.getComponentCpuQuotaPath(component.getServiceName()),
+                        cpuQuotaUsStr.getBytes(StandardCharsets.UTF_8));
+            }
+
+        } catch (IOException e) {
+            logger.atError().setCause(e).kv(COMPONENT_NAME, component.getServiceName())
+                    .log("Failed to apply resource limits");
+        }
+    }
+
+    @Override
+    public void resetResourceLimits(GenericExternalService component) {
+        for (Cgroup cg : ENABLED_CGROUPS) {
+            try {
+                Files.deleteIfExists(cg.getSubsystemComponentPath(component.getServiceName()));
+                Files.createDirectory(cg.getSubsystemComponentPath(component.getServiceName()));
+            } catch (IOException e) {
+                logger.atError().setCause(e).kv(COMPONENT_NAME, component.getServiceName())
+                        .log("Failed to remove the resource controller");
+            }
+        }
+    }
+
+    @Override
+    public void addComponentProcess(GenericExternalService component, Process process) {
+
+        if (!Files.exists(Cgroup.CPU.getSubsystemComponentPath(component.getServiceName()))
+                || !Files.exists(Cgroup.Memory.getSubsystemComponentPath(component.getServiceName()))) {
+            logger.atInfo().kv(COMPONENT_NAME, component.getServiceName()).log("Resource controller is not enabled");
+            return;
+        }
+
+        try {
+            Set<Integer> childProcesses = platform.getChildPids(process);
+            for (Integer pid : childProcesses) {
+                if (pid == null) {
+                    logger.atError().log("The process doesn't exist and is skipped");
+                    continue;
+                }
+
+                Files.write(Cgroup.Memory.getCgroupProcsPath(component.getServiceName()),
```

Comment:
does this actually work? Files.write will overwrite the existing file and not use append, so if there are multiple processes will this work correctly to add them all to the cgroup?

Reply:
I tested it and it's working.